packages/plugin-kafka/src/kafka.ts

```diff
@@ -0,0 +1,149 @@
+
+import {
+  TopicConfig,
+  ProducerStreamOptions,
+  ConsumerStreamOptions,
+  KafkaConfig,
+} from './types'
+
+const debug = debugLog('plugin-kafka')
+
+/**
+ * Relative priority inside the same plugin group type
+ */
+export const priority = 0
+export const name = 'kafka'
+export const type = PluginTypes.transport
+
+/**
+ * Defines service extension
+ */
+export interface KafkaPlugin {
+  rdKafkaConfig: KafkaConfig
+  createConsumerStream: (
+    streamOptions: ConsumerStreamOptions,
+    conf?: KafkaConfig,
+    topicConf?: TopicConfig
+  ) => Promise<ConsumerStream>
+  createProducerStream: (
+    streamOptions: ProducerStreamOptions,
+    conf?: KafkaConfig,
+    topicConf?: TopicConfig
+  ) => Promise<ProducerStream>
+  close: () => Promise<void>
+}
+
+type AnyStream = ConsumerStream | ProducerStream
+
+export class KafkaFactory implements KafkaPlugin {
+  rdKafkaConfig: KafkaConfig
+  private _streams: Set<AnyStream>
+
+  constructor(config: KafkaConfig) {
+    this.rdKafkaConfig = config
+    this._streams = new Set<AnyStream>()
+  }
+
+  async connectStream(stream: AnyStream) {
+    const client = Object.prototype.hasOwnProperty.call(stream, 'consumer')
+    ? (stream as ConsumerStream).consumer
+    : (stream as ProducerStream).producer
+
+    const connectError = once(stream, 'error')
+    const connectSuccess = once(client, 'ready')
+
+    const [raceResult] = await Promise.race([connectError, connectSuccess])
+
+    if (raceResult instanceof Error) {
+      throw raceResult
+    }
+
+    // We don't need this listener further
+    const [errorListener] = stream.listeners('error')
+    stream.removeListener('error', errorListener as (...args: any[]) => void)
+
+    // Kafka debugging
+    client.on('event.log', (log: any) => {
+      debug(`[${log.severity}](${log.fac}): ${log.message}`)
+    })
+
+    stream.on('close', () => {
+      if (this._streams.has(stream)) {
+        this._streams.delete(stream)
+      }
+    })
+
+    this._streams.add(stream)
+  }
+
+  async createConsumerStream(
+    opts: ConsumerStreamOptions, conf?: Partial<KafkaConfig>, topicConf?: TopicConfig
+  ): Promise<ConsumerStream> {
+    const consumerConfig = { ...this.rdKafkaConfig, ...conf }
+
+    const stream = createReadStream(consumerConfig, topicConf, opts)
+    await this.connectStream(stream)
+
+    return stream
+  }
+
+  async createProducerStream(
+    opts: ProducerStreamOptions, conf?: Partial<KafkaConfig>, topicConf?: TopicConfig
+  ): Promise<ProducerStream> {
+    const producerConfig = { ...this.rdKafkaConfig, ...conf }
+
+    const stream = createWriteStream(producerConfig, topicConf, opts)
+    await this.connectStream(stream)
+
+    return stream
+  }
+
+  async close(this: Microfleet) {
+    const kafka = this[name]
+    const streamsToClean = [...kafka._streams.values()]
+      .map((stream: AnyStream) => {
+        return promisify(stream.close.bind(stream))()
+      })
+    await Promise.all(streamsToClean)
```

Comment:
can we promisify client constructor so that we dont need to promisify each method? promisifying prototype would be more efficient

Reply:
fixed